src/Juvix/Compiler/Concrete/Translation/FromParsed/Analysis/Scoping.hs

```diff
@@ -2282,6 +2294,60 @@ resolveSyntaxDef = \case
   SyntaxOperator opDef -> resolveOperatorSyntaxDef opDef
   SyntaxIterator iterDef -> resolveIteratorSyntaxDef iterDef
 
+-------------------------------------------------------------------------------
+-- Check precedences are comparable
+-------------------------------------------------------------------------------
+
+checkPrecedences ::
+  forall r.
+  Members '[Error ScoperError, InfoTableBuilder] r =>
+  [S.Name] ->
+  Sem r ()
+checkPrecedences opers = do
+  tab <- getInfoTable
+  let fids = mapMaybe (^. fixityId) $ mapMaybe (^. S.nameFixity) opers
+      deps = createDependencyInfo (tab ^. infoPrecedenceGraph) mempty
+  mapM_ (uncurry (checkPath deps)) $
+    [(fid1, fid2) | fid1 <- fids, fid2 <- fids, fid1 /= fid2]
+  where
+    checkPath :: DependencyInfo S.NameId -> S.NameId -> S.NameId -> Sem r ()
+    checkPath deps fid1 fid2 =
+      unless (isPath deps fid1 fid2 || isPath deps fid2 fid1) $
+        throw (ErrIncomparablePrecedences (IncomaprablePrecedences (findOper fid1) (findOper fid2)))
+
+    findOper :: S.NameId -> S.Name
+    findOper fid =
+      fromJust $
+        find
+          (maybe False (\fx -> Just fid == (fx ^. fixityId)) . (^. S.nameFixity))
+          opers
+
+checkExpressionPrecedences :: Members '[Error ScoperError, InfoTableBuilder] r => ExpressionAtoms 'Scoped -> Sem r ()
+checkExpressionPrecedences (ExpressionAtoms atoms _) =
+  checkPrecedences opers
+  where
+    opers :: [S.Name]
+    opers = mapMaybe getIden (toList atoms)
+
+    getIden :: ExpressionAtom 'Scoped -> Maybe S.Name
+    getIden a = case a of
+      AtomIdentifier nm -> Just (nm ^. scopedIden)
+      _ -> Nothing
+
+checkPatternPrecedences :: Members '[Error ScoperError, InfoTableBuilder] r => PatternAtoms 'Scoped -> Sem r ()
+checkPatternPrecedences (PatternAtoms atoms _) =
+  checkPrecedences opers
+  where
+    opers :: [S.Name]
+    opers = mapMaybe getIden (toList atoms)
+
+    getIden :: PatternAtom 'Scoped -> Maybe S.Name
+    getIden = \case
+      PatternAtomIden i -> case i of
+        PatternScopedConstructor c -> Just c
+        _ -> Nothing
+      _ -> Nothing
```

Comment:
Move this to `src/Juvix/Compiler/Concrete/Extra.hs`

Reply:
done